fl4health/clients/adaptive_drift_constraint_client.py

```diff
@@ -0,0 +1,214 @@
+            return packed_params
+
+    def set_parameters(self, parameters: NDArrays, config: Config, fitting_round: bool) -> None:
+        """
+        Assumes that the parameters being passed contain model parameters concatenated with a penalty weight. They are
+        unpacked for the clients to use in training. In the first fitting round, we assume the full model is being
+        initialized and use the FullParameterExchanger() to set all model weights.
+        Args:
+            parameters (NDArrays): Parameters have information about model state to be added to the relevant client
+                model and also the penalty weight to be applied during training.
+            config (Config): The config is sent by the FL server to allow for customization in the function if desired.
+            fitting_round (bool): Boolean that indicates whether the current federated learning round is a fitting
+                round or an evaluation round. This is used to help determine which parameter exchange should be used
+                for pulling parameters. A full parameter exchanger is always used if the current federated learning
+                round is the very first fitting round.
+        """
+        assert self.model is not None and self.parameter_exchanger is not None
+
+        server_model_state, self.drift_penalty_weight = self.parameter_exchanger.unpack_parameters(parameters)
+        log(INFO, f"Penalty weight received from the server: {self.drift_penalty_weight}")
+
+        super().set_parameters(server_model_state, config, fitting_round)
+
+    def compute_training_loss(
+        self,
+        preds: TorchPredType,
+        features: TorchFeatureType,
+        target: TorchTargetType,
+    ) -> TrainingLosses:
+        """
+        Computes training loss given predictions of the model and ground truth data. Adds to objective by including
+        penalty loss.
+
+        Args:
+            preds (TorchPredType): Prediction(s) of the model(s) indexed by name. All predictions included in
+                dictionary will be used to compute metrics.
+            features: (TorchFeatureType): Feature(s) of the model(s) indexed by name.
+            target: (TorchTargetType): Ground truth data to evaluate predictions against.
+
+        Returns:
+            TrainingLosses: an instance of TrainingLosses containing backward loss and additional losses indexed
+                by name. Additional losses includes penalty loss.
+        """
+        loss, additional_losses = self.compute_loss_and_additional_losses(preds, features, target)
+        if additional_losses is None:
+            additional_losses = {}
+
+        additional_losses["loss"] = loss.clone()
+        # adding the vanilla loss to the additional losses to be used by update_after_train for potential adaptation
+        additional_losses["adaptation_loss"] = loss.clone()
```

Comment:
nit: Adaptation loss is a bit confusing to me. It seems like it could mean either the loss that we are using to adapt the constant applied to the weight drift loss or the weight drift loss which is adapted based on the constant from the server. I kind of liked criterion loss or loss without penalty as you described it in the init docstring

Reply:
My only hesitation with calling it criterion loss or loss without penalty is that it doesn't _necessarily_ have to be that. Do you think "loss_for_adaptation" is a bit better?

Reply:
Yeah, i think that is a bit more clear!